src/hotspot/share/runtime/threads.cpp

```diff
@@ -1417,9 +1419,30 @@ JavaThread *Threads::owning_thread_from_monitor_owner(ThreadsList * t_list,
   return the_owner;
 }
 
+JavaThread* Threads::owning_thread_from_object(ThreadsList * t_list, oop obj) {
+  assert(SafepointSynchronize::is_at_safepoint(), "not safe outside of safepoint");
```

Comment:
`ObjectSynchronizer::get_lock_owner()` calls `Threads::owning_thread_from_object()`
and I don't think you can assert that `ObjectSynchronizer::get_lock_owner()`  is only
called from a safepoint. In particular, `ThreadSnapshot::initialize()` calls
`ObjectSynchronizer::get_lock_owner()`  and I believe that `initialize()` function can
be called from non-safepoint places with M&M APIs...

Reply:
Update: In v35 you put back the `!UseFastLocking` check in `jmm_GetThreadInfo()`
so now the `assert()` won't fire, but you've again changed the behavior of that API
and now it will be able to observe fewer thread state changes than it did before.

Reply:
Please explain why you think this is "not safe". Yes, you can observe state that is in
the process of changing, but do you think that we'll see a crash with allowing
`Threads::owning_thread_from_object()` to be called from a non-safepoint place?

Reply:
I don't think we'd see a crash, but we might get false results when we are scanning the lock-stack of a foreign thread, when that thread does not hold still. I'm not even comfortable doing that cross-stack lock query with the old code.

Reply:
Given the owner could release the monitor the moment after we check I don't see how false results are an issue here. The existing code should be safe when not executed at a safepoint..